### ISSUE TITLE
Use context manager for temporary files during pytests

### ISSUE THREAD

**Issue by ewels** [nf-core/tools#1308]
I've just noticed that other tests write cache in `$HOME/.nextflow/nf-core`, while only the one I disabled raised issues. Personally, I don't like writing my cache during tests, however this cache seems to speed up testing process: maybe this fix will pristine the testing environment, but would be nice to cache this tests in a different temporary directory?

There's another point I noticed: all temporary folders are never removed since the user of `tempfile.mkdtemp` is [responsible for deleting the temporary directory and its contents](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp). Should this temporary folders be removed (for example with *context managers*):

```python
# create a temporary directory using the context manager
with tempfile.TemporaryDirectory() as tmpdirname:
    # do a lot of tests

# outside the with block, temporary folder is removed
```

_Originally posted by @bunop in https://github.com/nf-core/tools/issues/1296#issuecomment-949790577_

**Comment by bunop** [comment]
While digging in to code to propose a PR, I understood that is not always the best solution. There are test cases in which a temporary file (or directory) is generated in `setUp` and required in the following test: a `with` statement in `setUp` will close and erase the temporary files before the required test is run. In such cases I preferred to track the temporary folder (or file) and to delete it in `tearDown` step. The other thing I understood is that using a with statement in a test will make harder the revision process: since all test steps need to be indented in the `with` statement, you will have a lot of changes to revise even if I let code untouched: for such reason I chose to implement the context manager with decorators, and to decorate tests when possible. The last thing I noticed is there are also other points outside test in which temporary files are created, however my opinion was only to affect the test code since I don't want to affect both tests and code and I don't want to remove stuff when calling *nf-core tools* as a user. However, I removed the temporary folder in `nf_core/lint/files_unchanged.py` since it seems to me that a function is called during test (and this let 0 temporary files after tests): feel free to reject or undo this modification.

Now the only temporary files produced are generated by the pytest suite. Honestly, I don't know how to remove it, since I think they are produced by the `pytest` or `pytest-dataset`. If I find a solutions anywhere I can do a proposal

By the way, there are a lot of changes. Feel free to ask me questions or reject the PR if you don't like those changes
